server/client_server.go

```diff
@@ -0,0 +1,113 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+
+	bitbucketv1 "github.com/gfleury/go-bitbucket-v1"
+	"github.com/pkg/errors"
+	"golang.org/x/oauth2"
+)
+
+// TODO: This will be changed to create the main structs when modularized
+type Link struct {
+	Href string `json:"href"`
+}
+
+type BitbucketUser struct {
+	AccountId int    `json:"id"`
+	Username  string `json:"name"`
+	Links     struct {
+		Self []Link `json:"self"`
+	} `json:"links"`
+}
+
+type BitbucketServerClient struct {
+	apiClient *bitbucketv1.APIClient
+
+	selfHostedURL    string
+	selfHostedAPIURL string
+
+	token oauth2.Token
+}
+
+func NewClientServer() *BitbucketServerClient {
+	return &BitbucketServerClient{}
+}
+
+// TODO: This method needs to be changed when Modularization is built
+func (c *BitbucketServerClient) Connect(config Configuration, token oauth2.Token, ts oauth2.TokenSource) *BitbucketServerClient {
+	// setup Oauth context
+	auth := context.WithValue(context.Background(), bitbucketv1.ContextOAuth2, ts)
+
+	tc := oauth2.NewClient(auth, ts)
+
+	// create config for bitbucket API
+	configBb := bitbucketv1.NewConfiguration(config.BitbucketAPISelfHostedURL)
+	configBb.HTTPClient = tc
+
+	c.apiClient = bitbucketv1.NewAPIClient(context.Background(), configBb)
+	c.selfHostedURL = config.BitbucketSelfHostedURL
+	c.selfHostedAPIURL = config.BitbucketAPISelfHostedURL
+	c.token = token
+
+	return c
+}
+
+func (c *BitbucketServerClient) getWhoAmI() (string, error) {
+	baseUrl := fmt.Sprintf("%s/plugins/servlet/applinks/whoami", c.selfHostedURL)
+
+	req, err := http.NewRequest("GET", baseUrl, nil)
+	if err != nil {
+		return "", err
+	}
+
+	client := &http.Client{}
+
+	req.Header.Add("Authorization", fmt.Sprintf("Bearer %s", c.token.AccessToken))
```

Comment:
Here it is the `oauth2.Token`:

```
{
  "access_token": "eyJhbGciOiJIUzI1NiJ9.eyJpZCI6ImU5NDlhYzY5MDM5YWI5N2YyZWYyYjdhMzYwZjdlYTgzIn0.EJHNLcUShky7YJE8VKx5P3PiVHFbfoUmStjaTSrlq1g",
  "token_type": "bearer",
  "refresh_token": "eyJhbGciOiJIUzI1NiJ9.eyJpZCI6IjU2MWFmYzQ0OWVkYmE5MzcwMDUxMTU0NWYzZGI3ZTgzIn0.RPC5PLOsYzNpjBjxVRwpP5khyeCNDE7SFLnmZS6pQ3w",
  "expiry": "2024-01-09T11:22:01.072173-03:00"
}
```

I sent you a private message on Mattermost, but I am sending it here as well for Documentation.

We could replace it with the `oauth2.Client` then, it will probably handle the refreshment of the token by itself.

What do you think?

Reply:
as discussed, changed the Client to use the `oauth2.Client` one, thank you for that!

it's up to review.